File: mailwaycore/src/main/java/com/dimension/mailwaycore/data/entity/ChatMessage.kt
package com.dimension.mailwaycore.data.entity

import androidx.room.Embedded
import androidx.room.Entity
import androidx.room.PrimaryKey
import androidx.room.Relation
import com.fasterxml.jackson.annotation.JsonIgnore
import kotlinx.serialization.Serializable

@Entity
@Serializable
data class ChatMessage(
    @PrimaryKey
    @JsonIgnore
    var id: String,
    var message_id: String,
    var created_at: Long,
    var updated_at: Long,
    var message_timestamp: Long,
    var compose_timestamp: Long?,
    var receive_timestamp: Long,
    var share_timestamp: Long?,
    var sender_public_key: String,
    var recipient_public_keys: List<String>,
    var armored_message: String,
    var payload: String?,
    var payload_kind: PayloadKind,
    var version: Int?,
    @JsonIgnore
    var chatId: String,
    @JsonIgnore
    var quote_message_id: String?
)


@Serializable
data class ChatMessageAndQuoteMessage(
    @Embedded val chatMessage: ChatMessage,
    @Relation(
        parentColumn = "id",
        entityColumn = "chat_message_id"
    )
    val quoteMessage: QuoteMessage
)

File: mailway_ui/lib/data/entity/ChatMessage.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:mailwayui/data/entity/PayloadKind.dart';

part 'ChatMessage.g.dart';

@JsonSerializable()
class ChatMessage {
  String id;
  int created_at;
  int updated_at;
  int message_timestamp;
  int compose_timestamp;
  int receive_timestamp;
  int share_timestamp;
  String sender_public_key;
  List<String> recipient_public_keys;
  String armored_message;
  String payload;
  PayloadKind payload_kind;
  int version;
  String chatId;

  ChatMessage({
    this.id,
    this.created_at,
    this.updated_at,
    this.message_timestamp,
    this.compose_timestamp,
    this.receive_timestamp,
    this.share_timestamp,
    this.sender_public_key,
    this.recipient_public_keys,
    this.armored_message,
    this.payload,
    this.payload_kind,
    this.version,
    this.chatId,
  });
  
  factory ChatMessage.fromJson(Map<String, dynamic> json) =>
      _$ChatMessageFromJson(json);

  Map<String, dynamic> toJson() => _$ChatMessageToJson(this);
}
